src/explore-education-statistics-common/src/modules/table-tool/utils/applyTableHeadersOrder.ts

```diff
@@ -0,0 +1,439 @@
+  );
+  const reorderedGroupsOnOtherAxisOrder = reorderedGroupsOnOtherAxis.map(
+    group => getFilterType(group[0]),
+  );
+
+  const groupsAreReordered = reorderedGroupsOrder
+    ? areGroupsReordered(
+        reorderedGroupsOrder,
+        reorderedGroupsOnOtherAxisOrder,
+        updatedGroupsOrder,
+        updatedGroupsOnOtherAxisOrder,
+      )
+    : false;
+
+  const originalGroupsWereReordered = originalGroupsOrder
+    ? areGroupsReordered(
+        originalGroupsOrder,
+        originalGroupsOnOtherAxisOrder,
+        updatedGroupsOrder,
+        updatedGroupsOnOtherAxisOrder,
+      )
+    : false;
+
+  const additionalGroups =
+    reorderedGroupsOnOtherAxis && reorderedGroups
+      ? getAdditionalGroups(
+          reorderedGroupsOnOtherAxis,
+          reorderedGroups,
+          updatedGroups,
+        )
+      : getAdditionalGroups(
+          originalGroupsOnOtherAxis,
+          originalGroups,
+          updatedGroups,
+        );
+
+  // If the groups were reordered append new groups to the end, else maintain the default order.
+  return groupsAreReordered || originalGroupsWereReordered
+    ? [...orderedGroups, ...additionalGroups]
+    : [...orderedGroups, ...additionalGroups].sort((a, b) => {
+        return (
+          updatedGroupsOrder.indexOf(getFilterType(a[0])) -
+          updatedGroupsOrder.indexOf(getFilterType(b[0]))
+        );
+      });
+};
+
+/**
+ * Apply previous ordering, from the saved order and reordering, to updated table headers.
+ */
+const applyTableHeadersOrder = ({
+  originalTableHeaders,
+  reorderedTableHeaders,
+  updatedTableHeaders,
+}: {
+  originalTableHeaders?: TableHeadersConfig;
+  reorderedTableHeaders?: TableHeadersConfig;
+  updatedTableHeaders: TableHeadersConfig;
+}): TableHeadersConfig => {
+  if (!reorderedTableHeaders && !originalTableHeaders) {
+    return updatedTableHeaders;
+  }
+
+  const allOrderedColumnGroups = orderGroups({
+    axis: 'column',
+    originalTableHeaders,
+    reorderedTableHeaders,
+    updatedTableHeaders,
+  });
+
+  const allOrderedRowGroups = orderGroups({
+    axis: 'row',
+    originalTableHeaders,
+    reorderedTableHeaders,
+    updatedTableHeaders,
+  });
+
+  const allOrderedColumnGroupsLength = allOrderedColumnGroups.length;
+  const allOrderedRowGroupsLength = allOrderedRowGroups.length;
+
+  // For safety - can't think how this would be possible, but just in case, return the updated headers so the table doesn't break.
+  if (!allOrderedColumnGroupsLength && !allOrderedRowGroupsLength) {
+    return updatedTableHeaders;
+  }
+
+  // Make sure columns is populated by taking the first of the rows groups
+  if (!allOrderedColumnGroupsLength) {
+    return {
+      columnGroups: [],
+      columns: allOrderedRowGroups[0],
+      rowGroups: allOrderedRowGroups.filter(
+        (_, index) => index > 0 && index < allOrderedRowGroupsLength - 1,
+      ),
+      rows: allOrderedRowGroups[allOrderedRowGroupsLength - 1],
```

Comment:
Could use Lodash's `last`:

```ts
rows: last(allOrderedRowGroups)
```

Reply:
That causes a typescript error as the result of `last` can be undefined, even though `allOrderedRowGroups` is definitely populated at this point. So it'd need to be `last(allOrderedRowGroups) ?? []` which I think would be a bit misleading as rows can't be empty.

Reply:
Cool, fair enough!